subsys/bluetooth/mesh/proxy_cli.c

```diff
@@ -0,0 +1,314 @@
+
+	if (!memcmp(sub->keys[0].net_id, buf->data, 8)) {
+		return 1;
+	}
+
+	if (!bt_mesh_subnet_has_new_key(sub)) {
+		return 0;
+	}
+
+	if (!memcmp(sub->keys[1].net_id, buf->data, 8)) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static void proxy_msg_recv(struct bt_mesh_proxy_role *role)
+{
+	switch (role->msg_type) {
+	case BT_MESH_PROXY_NET_PDU:
+		BT_DBG("Mesh Network PDU");
+		bt_mesh_net_recv(&role->buf, 0, BT_MESH_NET_IF_PROXY);
+		break;
+	case BT_MESH_PROXY_BEACON:
+		BT_DBG("Mesh Beacon PDU");
+		bt_mesh_beacon_recv(&role->buf);
+		break;
+	case BT_MESH_PROXY_CONFIG:
+		BT_DBG("Mesh Configuration PDU");
+		/* TODO */
+		break;
+	default:
+		BT_WARN("Unhandled Message Type 0x%02x", role->msg_type);
+		break;
+	}
+}
+
+static void proxy_connected(struct bt_conn *conn)
+{
+	servers[bt_conn_index(conn)].srv =
+		bt_mesh_proxy_role_setup(conn, bt_mesh_gatt_send,
+					 proxy_msg_recv);
+
+}
+
+static void proxy_link_open(struct bt_conn *conn)
+{
+	servers[bt_conn_index(conn)].link_opened = true;
+}
+
+static void proxy_disconnected(struct bt_conn *conn)
+{
+	bt_mesh_proxy_role_cleanup(servers[bt_conn_index(conn)].srv);
+
+	servers[bt_conn_index(conn)].srv = NULL;
+	servers[bt_conn_index(conn)].link_opened = false;
+}
+
+static const struct bt_mesh_gatt proxy = {
+	.srv_uuid		= BT_UUID_INIT_16(BT_UUID_MESH_PROXY_VAL),
+	.data_in_uuid		= BT_UUID_INIT_16(BT_UUID_MESH_PROXY_DATA_IN_VAL),
+	.data_out_uuid		= BT_UUID_INIT_16(BT_UUID_MESH_PROXY_DATA_OUT_VAL),
+	.data_out_cccd_uuid	= BT_UUID_INIT_16(BT_UUID_GATT_CCC_VAL),
+
+	.connected		= proxy_connected,
+	.link_open		= proxy_link_open,
+	.disconnected		= proxy_disconnected
+};
+
+#if defined(CONFIG_BT_MESH_CDB)
+struct node_user_data {
+	const uint8_t *random;
+	const uint8_t *hash;
+	struct bt_mesh_cdb_node *node;
+};
+
+static uint8_t node_id_find(struct bt_mesh_cdb_node *node,
+			    void *user_data)
+{
+	int err;
+	uint8_t tmp[16];
+	const uint8_t *identity;
+	struct bt_mesh_subnet *sub;
+	struct node_user_data *ud = user_data;
+
+	if (!find_net_idx(node->net_idx)) {
+		return BT_MESH_CDB_ITER_CONTINUE;
+	}
+
+	(void)memset(tmp, 0, 6);
+	memcpy(tmp + 6, ud->random, 8);
+	sys_put_be16(node->addr, &tmp[14]);
+
+	sub = bt_mesh_subnet_get(node->net_idx);
+	if (sub == NULL) {
+		ud->node = NULL;
+		return BT_MESH_CDB_ITER_STOP;
+	}
+
+	identity = sub->keys[bt_mesh_subnet_has_new_key(sub)].identity;
+
+	err = bt_encrypt_be(identity, tmp, tmp);
+	if (err) {
+		ud->node = NULL;
+		return BT_MESH_CDB_ITER_STOP;
+	}
+
+	if (!memcmp(ud->hash, tmp + 8, 8)) {
+		ud->node = node;
+		return BT_MESH_CDB_ITER_STOP;
+	}
+
+	return BT_MESH_CDB_ITER_CONTINUE;
+}
+#endif
+
+void bt_mesh_proxy_adv_recv(const struct bt_le_scan_recv_info *info,
+			    struct net_buf_simple *buf)
+{
+	uint8_t type;
+	struct bt_mesh_subnet *sub;
+
+	BT_DBG("");
+
+	type = net_buf_simple_pull_u8(buf);
+	switch (type) {
+	case BT_MESH_ID_TYPE_NET:
+		sub = bt_mesh_subnet_find(subnet_find_cb, buf);
+		if (sub && find_net_idx(sub->net_idx)) {
```

Comment:
For performance, it'd be better to do this the other way around, so we only do subnet iteration if we're actively trying to connect to this network.

Reply:
Also, this needs to check whether we've already connected to this subnet